In [8]:
import sys
import ctypes    # It is an external module which can help us make the data types from the c language

In [3]:
L = []
for i in range(50):
    print(i,sys.getsizeof(L))
    L.append(i)

0 56
1 88
2 88
3 88
4 88
5 120
6 120
7 120
8 120
9 184
10 184
11 184
12 184
13 184
14 184
15 184
16 184
17 248
18 248
19 248
20 248
21 248
22 248
23 248
24 248
25 312
26 312
27 312
28 312
29 312
30 312
31 312
32 312
33 376
34 376
35 376
36 376
37 376
38 376
39 376
40 376
41 472
42 472
43 472
44 472
45 472
46 472
47 472
48 472
49 472


From above we can see that first the value of the size was 56 then 88 after 4 size later the size becomes 120 then after its double it becomes 184

and this goes ---> So we can conclude that the list act as a dynamic array however instead of doubling everytime in python we add **8** extra spaces

In [20]:


class MeraList: 
 def __init__(self):           # Creating a constructor
  self.size = 1                # size = Max amount of data which can be store
  self.n = 0                   # n = How much the space is filled currently

  # Creat a C type array with size = self.size

  self.A = self.__make__array(self.size)
 
 # Magic function to get the length of the array
 def __len__(self):
  return self.n
 
 # Append Function
 def append(self,item):
  if self.n == self.size:
   #resize 
   self.resize(self.size*2)

   # append
   self.A[self.n]=item
   self.n = self.n +1
  
 # Resize function 
 def __resize__(self , new_capacity):
  # Create a new array with new capacity
  B=self.__make_array(new_capacity)
  self.size = new_capacity
  # Copy the content of A to B
  for i in range(self.n):
   B[i]= self.A[i]
   # reassign A
   self.A = B 





 def __make__array(self,capacity):
  return (capacity*ctypes.py_object)()      # This code creates a Ctype array with size capacity ( static and referential array)




In [15]:
L = MeraList()

In [11]:
type(L)

__main__.MeraList

## Adding Length feature

This is done in the class itself by creating a magic function __len__(self) in the class and returning self.n as a return statement which gives current length of the array

In [12]:
len(L)

0

In [16]:
L = [1,2,3,4]

In [17]:
L

[1, 2, 3, 4]

## Creating the append feature
- Append : Adding the value at the last

Steps in creating the append feature is :
1. check whether the current size is full or not 
2. If not then we will just add that to the last
3. If it is full then the we make a new array of double the current size then copy the values of current array to the new one and add the new value to new array
